README.md
# Send invoices using Stripe

**Description**: Creates and sends brandable customer invoices using the Stripe payments platform



**Details**: **IMPORTANT: This extension is part of the Firebase Alpha program. [Alpha extensions](https://accounts.google.com/AccountChooser?service=gerritcodereview&continue=https://dev-partners.googlesource.com/login/samples/firebase/extensions-alpha/) are confidential, and their functionality might change in backward-incompatible ways before official, public release. We do not recommend installing Alpha extensions in production apps.**

Use this extension to create and send brandable customer invoices using the payments platform [Stripe](https://www.stripe.com/).

The invoices are automatically customized with the logo and color theme of your business that you've set up in Stripe. After the invoice is sent, you can use the Stripe dashboard to track whether the customer has paid and how much money you processed with detailed reporting and charts.

This extension listens to your specified Cloud Firestore collection for new documents (like the example below). When you add a document, Stripe uses the invoice information in the document to create an invoice in their system then sends the invoice to the email address specified in the document. You can optionally manage your customer email addresses using [Firebase Authentication](https://firebase.google.com/docs/auth) user IDs. 

```js
email: "customer@example.com",
items: [{
    amount: 2000,
    currency: "usd",
    description: "Growth plan"
}]
```

Here's an example of what your customized invoice will look like!

![An invoice page showing an itemized receipt, with options to pay with card or bank transfer](https://www.gstatic.com/mobilesdk/200421_mobilesdk/hosted-invoice-page.png)

An optional feature of this extension is to automatically update the invoice's status in its Cloud Firestore document. You can configure this feature after installing the extension by registering a Stripe webhook that listens for [Stripe invoice events](https://stripe.com/docs/api/events/types#event_types-invoice.created). If you want to use this optional feature, leave the parameter `Stripe webhook secret` empty during installation, then reconfigure your installed extension later with the actual value for your registered webhook. More details about this process are provided after installation.

#### Additional setup

Before installing this extension, set up the following Firebase services in your Firebase project:

* [Cloud Firestore](https://firebase.google.com/docs/firestore) to store invoice information and optionally invoice status
* [Firebase Authentication](https://firebase.google.com/docs/auth) to optionally manage email and customer data

You must also have a Stripe account and a [Stripe API key](https://dashboard.stripe.com/apikeys) before installing this extension.

**Note:** Stripe has a test mode that lets you make API calls without making actual payments. To use this extension with Stripe's test mode, set the extension's `Stripe API key` parameter (during extension configuration) to use a test mode key. A test mode key looks like `rk_test_12345`, whereas a live mode key would be `rk_live_12345`. As this extension only requires write access to your Stripe `customers` and `invoices` resources, we recommend that you create a [restricted key](https://stripe.com/docs/keys#limit-access) with limited access to only these resources rather than using your secret key.

#### Pricing

This extension uses the following Firebase services which may have associated charges:

* Cloud Firestore
* Cloud Functions
* Firebase Authentication (optional)

When you use Firebase Extensions, you're only charged for the underlying resources that you use. A paid-tier billing plan is only required if the extension uses a service that requires a paid-tier plan, for example calling to a Google Cloud Platform API or making outbound network requests to non-Google services. All Firebase services offer a free tier of usage. [Learn more about Firebase billing.](https://firebase.google.com/pricing)

Usage of this extension also requires you to have a Stripe account. You are responsible for any associated costs with your usage of Stripe and the [Stripe Billing product](https://stripe.com/pricing#billing-pricing).




**Configuration Parameters:**

* Cloud Functions deployment location: Where do you want to deploy the functions created for this extension? You usually want a location close to your database. For help selecting a location, refer to the [location selection guide](https://firebase.google.com/docs/functions/locations).

* Invoices collection: What is the path to the Cloud Firestore collection where you'll store your invoices?

* Stripe API key with restricted access: What is your Stripe API key for sending invoices?  We recommend creating a new [restricted key](https://stripe.com/docs/keys#limit-access) with write access only for the `customers` and `invoices` resources. If desired, you can optionally use a test mode API key for initial testing of your extension, but you'll need to later reconfigure your extension to use a live mode key. Learn more about API keys in your [Stripe dashboard](https://dashboard.stripe.com/apikeys).

* Days until payment is due: What is the default number of days the customer has before their payment is due? The invoice automatically closes after this number of days. You can override this default value for each invoice.

* Stripe webhook secret: This is your signing secret for a Stripe-registered webhook that updates each Cloud Firestore document with the invoice's status from the Stripe dashboard. This webhook can only be registered after installation. To use this optional feature, leave this value empty during installation, then follow the postinstall instructions for registering your webhook and configuring this value.



**Cloud Functions:**

* **sendInvoice:** Creates a Stripe Invoice when a new invoice document is created in your specified Cloud Firestore collection

* **updateInvoice:** If registered as a Stripe webhook, this function updates the appropriate invoice document in Cloud Firestore when a new Stripe Event is received.



**Access Required**:



This extension will operate with the following project IAM roles:

* firebaseauth.viewer (Reason: Allows the extension to find the email of a user based on their Firebase Authentication user ID.)

* datastore.user (Reason: Allows the extension to update invoices in Cloud Firestore with data from Stripe.)


CHANGELOG.md
## Version 0.1.1

[fixed] - Fixed idempotency key collision for invoice item creation which was causing errors for invoices with more than one item.

## Version 0.1.0

Initial release of the `firestore-invoice-stripe` extension.

functions/lib/logs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function startInvoiceCreate() {
    console.log("🙂 Received new invoice, starting processing");
}
exports.startInvoiceCreate = startInvoiceCreate;
function startInvoiceUpdate(eventType) {
    console.log(`🙂 Received new invoice event ${eventType}, starting processing`);
}
exports.startInvoiceUpdate = startInvoiceUpdate;
function error(err) {
    console.error("😞[Error] Unhandled error occurred during processing", err);
}
exports.error = error;
function missingPayload(payload) {
    if (!payload.items.length) {
        console.error(new Error("😞[Error] Missing at least one line item in items[]"));
    }
    if (!payload.email && !payload.uid) {
        console.error(new Error("😞[Error] Missing either a customer email address or Firebase Authentication uid"));
    }
}
exports.missingPayload = missingPayload;
function stripeError(err) {
    console.error("😞[Error] Error when making a request to the Stripe API:", err);
}
exports.stripeError = stripeError;
function invoiceCreatedError(invoice) {
    console.error(new Error("😞[Error] Error when creating the invoice:"), invoice);
}
exports.invoiceCreatedError = invoiceCreatedError;
function customerCreated(id, livemode) {
    console.log(`👤 Created a new customer: https://dashboard.stripe.com${livemode ? "" : "/test"}/customers/${id}`);
}
exports.customerCreated = customerCreated;
function customerRetrieved(id, livemode) {
    console.log(`🙋 Found existing customer by email: https://dashboard.stripe.com${livemode ? "" : "/test"}/customers/${id}`);
}
exports.customerRetrieved = customerRetrieved;
function invoiceCreated(id, livemode) {
    console.log(`🧾 Created invoice: https://dashboard.stripe.com${livemode ? "" : "/test"}/invoices/${id}`);
}
exports.invoiceCreated = invoiceCreated;
function invoiceSent(id, email, hostedInvoiceUrl) {
    console.log(`📧 Sent invoice ${id} to ${email}: ${hostedInvoiceUrl}`);
}
exports.invoiceSent = invoiceSent;
function badSignature(err) {
    console.error("😞[Error] Webhook signature verification failed. Is your Stripe webhook secret parameter configured correctly?", err);
}
exports.badSignature = badSignature;
function malformedEvent(event) {
    var _a, _b, _c;
    let err;
    if (!((_b = (_a = event) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.object)) {
        err = new Error("Could not find event.data.object");
    }
    else if (!((_c = event) === null || _c === void 0 ? void 0 : _c.type)) {
        err = new Error("Could not find event.type");
    }
    console.error("😞[Error] Malformed event", err);
}
exports.malformedEvent = malformedEvent;
function ignoreEvent(eventType) {
    console.log(`🙈 Ignoring event "${eventType}" because it because it isn't a relevant part of the invoice lifecycle`);
}
exports.ignoreEvent = ignoreEvent;
function unexpectedInvoiceAmount(numInvoices, invoiceId) {
    console.error("😞[Error] could not find invoice", new Error(`Expected 1 invoice with ID "${invoiceId}", but found ${numInvoices}`));
}
exports.unexpectedInvoiceAmount = unexpectedInvoiceAmount;
function statusUpdateComplete(invoiceId, newStatus, eventType) {
    console.log(`🙂 Updated invoice "${invoiceId}" to status "${newStatus}" on event type "${eventType}"`);
}
exports.statusUpdateComplete = statusUpdateComplete;
//# sourceMappingURL=logs.js.map

functions/src/logs.ts
import Stripe from "stripe";
import { InvoicePayload } from "./interfaces";

export function startInvoiceCreate() {
  console.log("🙂 Received new invoice, starting processing");
}

export function startInvoiceUpdate(eventType: string) {
  console.log(`🙂 Received new invoice event ${eventType}, starting processing`);
}

export function error(err: Error) {
  console.error("😞[Error] Unhandled error occurred during processing", err);
}

export function missingPayload(payload: InvoicePayload) {
  if (!payload.items.length) {
    console.error(
      new Error("😞[Error] Missing at least one line item in items[]")
    );
  }
  if (!payload.email && !payload.uid) {
    console.error(
      new Error(
        "😞[Error] Missing either a customer email address or Firebase Authentication uid"
      )
    );
  }
}

export function stripeError(err: Stripe.StripeCardError) {
  console.error(
    "😞[Error] Error when making a request to the Stripe API:",
    err
  );
}

export function invoiceCreatedError(invoice?: Stripe.Invoice) {
  console.error(
    new Error("😞[Error] Error when creating the invoice:"),
    invoice
  );
}

export function customerCreated(id: string, livemode: boolean) {
  console.log(
    `👤 Created a new customer: https://dashboard.stripe.com${livemode ? "" : "/test"}/customers/${id}`
  );
}

export function customerRetrieved(id: string, livemode: boolean) {
  console.log(
    `🙋 Found existing customer by email: https://dashboard.stripe.com${livemode ? "" : "/test"}/customers/${id}`
  );
}

export function invoiceCreated(id: string, livemode: boolean) {
  console.log(
    `🧾 Created invoice: https://dashboard.stripe.com${livemode ? "" : "/test"}/invoices/${id}`
  );
}

export function invoiceSent(
  id: string,
  email: string,
  hostedInvoiceUrl: string
) {
  console.log(`📧 Sent invoice ${id} to ${email}: ${hostedInvoiceUrl}`);
}

export function badSignature(err: Error) {
  console.error(
    "😞[Error] Webhook signature verification failed. Is your Stripe webhook secret parameter configured correctly?",
    err
  );
}

export function malformedEvent(event: Stripe.Event) {
  let err;

  if (!event?.data?.object) {
    err = new Error("Could not find event.data.object");
  } else if (!event?.type) {
    err = new Error("Could not find event.type");
  }

  console.error("😞[Error] Malformed event", err);
}

export function ignoreEvent(eventType: string) {
  console.log(
    `🙈 Ignoring event "${eventType}" because it because it isn't a relevant part of the invoice lifecycle`
  );
}

export function unexpectedInvoiceAmount(
  numInvoices: number,
  invoiceId: string
) {
  console.error(
    "😞[Error] could not find invoice",
    new Error(
      `Expected 1 invoice with ID "${invoiceId}", but found ${numInvoices}`
    )
  );
}

export function statusUpdateComplete(
  invoiceId: string,
  newStatus: string,
  eventType: string
) {
  console.log(
    `🙂 Updated invoice "${invoiceId}" to status "${newStatus}" on event type "${eventType}"`
  );
}


functions/src/index.ts
import * as admin from "firebase-admin";
import * as functions from "firebase-functions";
import Stripe from "stripe";
import { InvoicePayload, OrderItem } from "./interfaces";
import * as logs from "./logs";
import config from "./config";

const stripe = new Stripe(config.stripeSecretKey, {
  apiVersion: "2019-12-03",
  // Register extension as a Stripe plugin
  // https://stripe.com/docs/building-plugins#setappinfo
  appInfo: {
    name: "Firebase firestore-invoice-stripe",
    version: "0.1.1"
  }
});

admin.initializeApp();

/* Creates a new invoice using Stripe */
const createInvoice = async function(
  customer: Stripe.Customer,
  orderItems: Array<OrderItem>,
  daysUntilDue: number,
  idempotencyKey: string
) {
  try {
    // Create an invoice item for each item in the document
    const itemPromises: Array<Promise<Stripe.InvoiceItem>> = orderItems.map(
      (item, index) => {
        return stripe.invoiceItems.create(
          {
            customer: customer.id,
            amount: item.amount,
            currency: item.currency,
            description: item.description
          },
          { idempotencyKey: `invoiceItems-create-${idempotencyKey}-${index}` }
        );
      }
    );

    // Create the individual invoice items for this customer from the items in payload
    const items: Array<Stripe.InvoiceItem> = await Promise.all(itemPromises);

    const invoice: Stripe.Invoice = await stripe.invoices.create(
      {
        customer: customer.id,
        collection_method: "send_invoice",
        days_until_due: daysUntilDue,
        auto_advance: true
      },
      { idempotencyKey: `invoices-create-${idempotencyKey}` }
    );
    logs.invoiceCreated(invoice.id, invoice.livemode)
    return invoice;
  } catch (e) {
    logs.stripeError(e);
    return null;
  }
};

/* Emails an invoice to a customer when a new document is created */
export const sendInvoice = functions.handler.firestore.document.onCreate(
  async (snap, context) => {
    try {
      const payload = snap.data() as InvoicePayload;
      const daysUntilDue =
        payload.daysUntilDue || config.daysUntilDue;

      if (!(payload.email || payload.uid) || !payload.items.length) {
        logs.missingPayload(payload);
        return;
      }

      // Background functions fire "at least once"
      // https://firebase.google.com/docs/functions/locations#background_functions
      //
      // This event ID will be the same for the same Cloud Firestore write
      // Use this as an idempotency key when calling the Stripe API
      const eventId = context.eventId;

      logs.startInvoiceCreate();

      let email;

      if (payload.uid) {
        // Look up the Firebase Authentication UserRecord to get the email
        const user = await admin.auth().getUser(payload.uid);
        email = user.email;
      } else {
        // Use the email provided in the payload
        email = payload.email;
      }

      // Check to see if there's a Stripe customer associated with the email address
      let customers: Stripe.ApiList<Stripe.Customer> = await stripe.customers.list(
        { email: payload.email }
      );
      let customer: Stripe.Customer;

      if (customers.data.length) {
        // Use the existing customer
        customer = customers.data[0];
        logs.customerRetrieved(customer.id, customer.livemode);
      } else {
        // Create new Stripe customer with this email
        customer = await stripe.customers.create(
          {
            email,
            metadata: {
              createdBy: "Created by the Firebase Extension: Send Invoices using Stripe" // optional metadata, adds a note
            }
          },
          { idempotencyKey: `customers-create-${eventId}` }
        );

        logs.customerCreated(customer.id, customer.livemode);
      }

      const invoice: Stripe.Invoice = await createInvoice(
        customer,
        payload.items,
        daysUntilDue,
        eventId
      );

      if (invoice) {
        // Write the Stripe Invoice ID back to the document in Cloud Firestore
        // so that we can find it in the webhook
        await snap.ref.update({
          stripeInvoiceId: invoice.id,
          stripeInvoiceRecord: `https://dashboard.stripe.com${invoice.livemode ? "" : "/test"}/invoices/${invoice.id}`
        });

        // Email the invoice to the customer
        const result: Stripe.Invoice = await stripe.invoices.sendInvoice(
          invoice.id,
          { idempotencyKey: `invoices-sendInvoice-${eventId}` }
        );
        if (result.status === "open") {
          // Successfully emailed the invoice
          logs.invoiceSent(result.id, email, result.hosted_invoice_url);
        } else {
          logs.invoiceCreatedError(result);
        }
      } else {
        logs.invoiceCreatedError();
      }
    } catch (e) {
      logs.error(e);
    }
    return;
  }
);

const relevantInvoiceEvents = new Set([
  "invoice.created",
  "invoice.finalized",
  "invoice.payment_failed",
  "invoice.payment_succeeded",
  "invoice.payment_action_required",
  "invoice.voided",
  "invoice.marked_uncollectible"
]);

/* A Stripe webhook that updates each invoice's status in Cloud Firestore */
export const updateInvoice = functions.handler.https.onRequest(
  async (req: functions.https.Request, resp) => {
    let event: Stripe.Event;

    // Instead of getting the `Stripe.Event`
    // object directly from `req.body`,
    // use the Stripe webhooks API to make sure
    // this webhook call came from a trusted source
    try {
      event = stripe.webhooks.constructEvent(
        req.rawBody,
        req.headers["stripe-signature"],
        config.stripeWebhookSecret
      );
    } catch (err) {
      logs.badSignature(err);
      resp.status(401).send("Webhook Error: Invalid Secret");
      return;
    }

    let invoice: Stripe.Invoice;
    let eventType: string;

    try {
      invoice = event.data.object as Stripe.Invoice;
      eventType = event.type;
    } catch (err) {
      logs.malformedEvent(event);
      resp.status(400).send(`Webhook Error: ${err.message}`);
      return;
    }

    if (!relevantInvoiceEvents.has(eventType)) {
      logs.ignoreEvent(eventType);

      // Return a response to Stripe to acknowledge receipt of the event
      resp.json({ received: true });
      return;
    }

    logs.startInvoiceUpdate(eventType);

    let invoicesInFirestore = await admin
      .firestore()
      .collection(config.invoicesCollectionPath)
      .where("stripeInvoiceId", "==", invoice.id)
      .get();

    if (invoicesInFirestore.size !== 1) {
      logs.unexpectedInvoiceAmount(invoicesInFirestore.size, invoice.id);

      resp.status(500).send(`Invoice not found.`);
      return;
    }

    // Keep a special status for `payment_failed`
    // because otherwise the invoice would still be marked `open`
    const invoiceStatus =
      eventType === "invoice.payment_failed"
        ? "payment_failed"
        : invoice.status;

    const doc = invoicesInFirestore.docs[0];
    await doc.ref.update({
      stripeInvoiceStatus: invoiceStatus,
      lastStripeEvent: eventType
    });

    logs.statusUpdateComplete(invoice.id, invoiceStatus, eventType);

    // Return a response to Stripe to acknowledge receipt of the event
    resp.json({ received: true });
  }
);


functions/lib/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const admin = __importStar(require("firebase-admin"));
const functions = __importStar(require("firebase-functions"));
const stripe_1 = __importDefault(require("stripe"));
const logs = __importStar(require("./logs"));
const config_1 = __importDefault(require("./config"));
const stripe = new stripe_1.default(config_1.default.stripeSecretKey, {
    apiVersion: "2019-12-03",
    // Register extension as a Stripe plugin
    // https://stripe.com/docs/building-plugins#setappinfo
    appInfo: {
        name: "Firebase firestore-invoice-stripe",
        version: "0.1.1"
    }
});
admin.initializeApp();
/* Creates a new invoice using Stripe */
const createInvoice = async function (customer, orderItems, daysUntilDue, idempotencyKey) {
    try {
        // Create an invoice item for each item in the document
        const itemPromises = orderItems.map((item, index) => {
            return stripe.invoiceItems.create({
                customer: customer.id,
                amount: item.amount,
                currency: item.currency,
                description: item.description
            }, { idempotencyKey: `invoiceItems-create-${idempotencyKey}-${index}` });
        });
        // Create the individual invoice items for this customer from the items in payload
        const items = await Promise.all(itemPromises);
        const invoice = await stripe.invoices.create({
            customer: customer.id,
            collection_method: "send_invoice",
            days_until_due: daysUntilDue,
            auto_advance: true
        }, { idempotencyKey: `invoices-create-${idempotencyKey}` });
        logs.invoiceCreated(invoice.id, invoice.livemode);
        return invoice;
    }
    catch (e) {
        logs.stripeError(e);
        return null;
    }
};
/* Emails an invoice to a customer when a new document is created */
exports.sendInvoice = functions.handler.firestore.document.onCreate(async (snap, context) => {
    try {
        const payload = snap.data();
        const daysUntilDue = payload.daysUntilDue || config_1.default.daysUntilDue;
        if (!(payload.email || payload.uid) || !payload.items.length) {
            logs.missingPayload(payload);
            return;
        }
        // Background functions fire "at least once"
        // https://firebase.google.com/docs/functions/locations#background_functions
        //
        // This event ID will be the same for the same Cloud Firestore write
        // Use this as an idempotency key when calling the Stripe API
        const eventId = context.eventId;
        logs.startInvoiceCreate();
        let email;
        if (payload.uid) {
            // Look up the Firebase Authentication UserRecord to get the email
            const user = await admin.auth().getUser(payload.uid);
            email = user.email;
        }
        else {
            // Use the email provided in the payload
            email = payload.email;
        }
        // Check to see if there's a Stripe customer associated with the email address
        let customers = await stripe.customers.list({ email: payload.email });
        let customer;
        if (customers.data.length) {
            // Use the existing customer
            customer = customers.data[0];
            logs.customerRetrieved(customer.id, customer.livemode);
        }
        else {
            // Create new Stripe customer with this email
            customer = await stripe.customers.create({
                email,
                metadata: {
                    createdBy: "Created by the Firebase Extension: Send Invoices using Stripe" // optional metadata, adds a note
                }
            }, { idempotencyKey: `customers-create-${eventId}` });
            logs.customerCreated(customer.id, customer.livemode);
        }
        const invoice = await createInvoice(customer, payload.items, daysUntilDue, eventId);
        if (invoice) {
            // Write the Stripe Invoice ID back to the document in Cloud Firestore
            // so that we can find it in the webhook
            await snap.ref.update({
                stripeInvoiceId: invoice.id,
                stripeInvoiceRecord: `https://dashboard.stripe.com${invoice.livemode ? "" : "/test"}/invoices/${invoice.id}`
            });
            // Email the invoice to the customer
            const result = await stripe.invoices.sendInvoice(invoice.id, { idempotencyKey: `invoices-sendInvoice-${eventId}` });
            if (result.status === "open") {
                // Successfully emailed the invoice
                logs.invoiceSent(result.id, email, result.hosted_invoice_url);
            }
            else {
                logs.invoiceCreatedError(result);
            }
        }
        else {
            logs.invoiceCreatedError();
        }
    }
    catch (e) {
        logs.error(e);
    }
    return;
});
const relevantInvoiceEvents = new Set([
    "invoice.created",
    "invoice.finalized",
    "invoice.payment_failed",
    "invoice.payment_succeeded",
    "invoice.payment_action_required",
    "invoice.voided",
    "invoice.marked_uncollectible"
]);
/* A Stripe webhook that updates each invoice's status in Cloud Firestore */
exports.updateInvoice = functions.handler.https.onRequest(async (req, resp) => {
    let event;
    // Instead of getting the `Stripe.Event`
    // object directly from `req.body`,
    // use the Stripe webhooks API to make sure
    // this webhook call came from a trusted source
    try {
        event = stripe.webhooks.constructEvent(req.rawBody, req.headers["stripe-signature"], config_1.default.stripeWebhookSecret);
    }
    catch (err) {
        logs.badSignature(err);
        resp.status(401).send("Webhook Error: Invalid Secret");
        return;
    }
    let invoice;
    let eventType;
    try {
        invoice = event.data.object;
        eventType = event.type;
    }
    catch (err) {
        logs.malformedEvent(event);
        resp.status(400).send(`Webhook Error: ${err.message}`);
        return;
    }
    if (!relevantInvoiceEvents.has(eventType)) {
        logs.ignoreEvent(eventType);
        // Return a response to Stripe to acknowledge receipt of the event
        resp.json({ received: true });
        return;
    }
    logs.startInvoiceUpdate(eventType);
    let invoicesInFirestore = await admin
        .firestore()
        .collection(config_1.default.invoicesCollectionPath)
        .where("stripeInvoiceId", "==", invoice.id)
        .get();
    if (invoicesInFirestore.size !== 1) {
        logs.unexpectedInvoiceAmount(invoicesInFirestore.size, invoice.id);
        resp.status(500).send(`Invoice not found.`);
        return;
    }
    // Keep a special status for `payment_failed`
    // because otherwise the invoice would still be marked `open`
    const invoiceStatus = eventType === "invoice.payment_failed"
        ? "payment_failed"
        : invoice.status;
    const doc = invoicesInFirestore.docs[0];
    await doc.ref.update({
        stripeInvoiceStatus: invoiceStatus,
        lastStripeEvent: eventType
    });
    logs.statusUpdateComplete(invoice.id, invoiceStatus, eventType);
    // Return a response to Stripe to acknowledge receipt of the event
    resp.json({ received: true });
});
//# sourceMappingURL=index.js.map